### PR TITLE
persist: add a test to catch changes in the end-to-end encoding

### PR DESCRIPTION
**Pull request by danhhz**
In the short-term, we are persisting only data that can safely be wiped
in a pinch (specifically the mz_metrics and mz_metric_histogram system
tables). At the very root of the metadata and data stored by persist,
and the first thing read on startup, is BlobMeta. It's encoded with a
version (BlobMeta::CURRENT_VERSION) and stored in blob storage. Our
short-term strategy (~1-2 months) to deal with backward compatibility is
to bump this version when any of the encodings used by persist change.
If persist starts up and encounters a BlobMeta encoding version less
than its current one, it's safe to simply wipe it and start fresh. (If
it encounters one that's greater, it should error out, because that
could be from a version of materialize where we _are_ persisting
un-wipe-able data.)

For this to work, we need a robust way to catch changes that indicate
that a BlobMeta encoding version bump is needed. Code review will catch
some of them, and this test will catch more of them. It will have false
positives, which is absolutely fine (a manual verification that we don't
need to bump the version can be declared by updating the "golden"), but
it hopefully will not have any false negatives.

Reminder that the real answer here is to make persist transparently
handle data written by previous versions of persist. It's a solvable
problem, we just haven't gotten there yet. Once we do, this test will be
modified to be a series of tests that verify we do correctly handle old
versions of persist data.

I manually verified that this test catches changes in the encoding of
BlobMeta and BlobTraceBatch. Also verified that it's not flaky using
https://crates.io/crates/cargo-stress.

### CODE REVIEW

File: src/persist/src/golden_test.rs [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+}
+
+#[derive(Debug, PartialEq, Eq)]
+struct PersistState {
+    streams: Vec<PersistStreamState>,
+}
+
+impl PersistState {
+    fn slurp_from(persist: &RuntimeClient) -> Result<Self, Error> {
+        // TODO: There is currently no way to get the set of active streams out
```
Comment (aljoscha) [inline src/persist/src/golden_test.rs:201]: I imagine this is a TODO for the future, not for this PR?

**Review by aljoscha — APPROVED** [review]
I  couldn't find anything that's obviously wrong. Plus, I like it a lot that we automate tests that should be automated/can catch inadvertent changes to the format. :tada:

File: src/persist/src/golden_test.rs (hunk 1 above)
Comment (danhhz) [reply src/persist/src/golden_test.rs:201]: Exactly, yes. I personally use TODO for things to fix later and WIP for things to fix before merging.

**Review by danhhz — COMMENTED** [review]
Thanks for the review!

**Pull request merged**